src/dcm/agent/ossec.py

```diff
@@ -0,0 +1,104 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from collections import namedtuple
+from threading import Thread
+from queue import Queue
+import socket
+import struct
+import json
+import hashlib
+
+OssecAlert = namedtuple('OssecAlert', 'alert_time alert_hash level rule message')
+
+class OssecSyslogUdpListener(object):
+
+    def __init__(self, sender, port=2003, ipaddr='127.0.0.1'):
+        """
+
+        :param sender:  an object that lets you send out alerts as they come in.
+        it will have the method "send" and it will take an OssecAlert (defined
+        elsewhere) object
+        :param port:  the udp port to listen on.  this will come from config
+        :param ipaddr:  the ip address to listen on.  this will almost always
+        be localhost
+        :return:
+        """
+        self.sender = sender
+        self.port = port
+        self.ipaddr = ipaddr
+        self.shutting_down = None
+        self.q = Queue(maxsize=-1)
+
+    def send_alert(self):
+        while not self.q.empty():
+           self.sender.send(self.q.get_nowait())
+
+    def start(self):
+        """
+        Start the udp listener thread
+        :return:
+
+        """
+        self.shutting_down = False
+        sock = self._build_socket()
+        t = Thread(target=self._listen_udp,
+                   args=(sock,),
+                   name="UDPListener")
+        t.start()
+        t.join()
+
+    def stop(self):
+        """"
+        Stop the udp listener thread
+        """
+
+    def _listen_udp(self, sock):
+        while not self.shutting_down:
+            data = sock.recv()
+            self._parse_data(data)
+
+    def _build_socket(self):
+        sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM, socket.IPPROTO_ROUTING)
+        sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
+        sock.bind(('', self.port))
+        mreq = struct.pack("=4sl", socket.inet_aton(self.ipaddr), socket.INADDR_ANY)
+        sock.setsockopt(socket.IPPROTO_IP, socket.IP_ADD_MEMBERSHIP, mreq)
+        return sock
+
+    def _build_response(self, d):
+        alert_time = d['message']
+        m = hashlib.md5()
+        m.update(d)
+        alert_hash = m.digest()
+        level = d['crit']
+        rule = d['classification']
+        message = d['message']
+        return alert_time, alert_hash, level, rule, message
+
+    def _parse_data(self, data):
+        start = data.find("{")
+        while start >= 0:
+            end = data.find("}")
+            jdoc = data[start, end+1]
+            d = json.loads(jdoc)
+            alert_res = self._build_response(d)
+            alert_tuple = OssecAlert(alert_res[0],
+                                     alert_res[1],
+                                     alert_res[2],
+                                     alert_res[3],
+                                     alert_res[4])
+            self.q.put_nowait(alert_tuple)
```

Comment:
so here instead of adding the parsed event to a queue, just directly send it to dcm.